File: ios/MossyUiModule.swift
import ExpoModulesCore
import ExpoUI
import SwiftUI

public class MossyUiModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MossyUi")

    OnCreate {
      ViewModifierRegistry.register("mossySizeConstraints") { params, appContext, _ in
        try MossySizeConstraintsModifier(from: params, appContext: appContext)
      }
      ViewModifierRegistry.register("mossyLinearGradientBackground") { params, appContext, _ in
        try MossyLinearGradientBackgroundModifier(from: params, appContext: appContext)
      }
      ViewModifierRegistry.register("mossyUnevenCornerRadius") { params, appContext, _ in
        try MossyUnevenCornerRadiusModifier(from: params, appContext: appContext)
      }
      ViewModifierRegistry.register("mossyDirectionalBorder") { params, appContext, _ in
        try MossyDirectionalBorderModifier(from: params, appContext: appContext)
      }
    }

    OnDestroy {
      ViewModifierRegistry.unregister("mossySizeConstraints")
      ViewModifierRegistry.unregister("mossyLinearGradientBackground")
      ViewModifierRegistry.unregister("mossyUnevenCornerRadius")
      ViewModifierRegistry.unregister("mossyDirectionalBorder")
    }
  }
}

struct MossySizeConstraintsModifier: ViewModifier, Record {
  @Field var minWidth: CGFloat?
  @Field var maxWidth: CGFloat?
  @Field var minHeight: CGFloat?
  @Field var maxHeight: CGFloat?

  func body(content: Content) -> some View {
    content.frame(
      minWidth: minWidth,
      maxWidth: maxWidth,
      minHeight: minHeight,
      maxHeight: maxHeight
    )
  }
}

struct MossyGradientPoint: Record {
  @Field var x: CGFloat = 0
  @Field var y: CGFloat = 0
}

struct MossyLinearGradientBackgroundModifier: ViewModifier, Record {
  @Field var colors: [Color] = []
  @Field var locations: [Double] = []
  @Field var startPoint: MossyGradientPoint = MossyGradientPoint()
  @Field var endPoint: MossyGradientPoint = MossyGradientPoint()

  func body(content: Content) -> some View {
    if colors.isEmpty {
      content
    } else {
      content.background(
        LinearGradient(
          stops: gradientStops,
          startPoint: UnitPoint(x: startPoint.x, y: startPoint.y),
          endPoint: UnitPoint(x: endPoint.x, y: endPoint.y)
        )
      )
    }
  }

  private var gradientStops: [Gradient.Stop] {
    colors.enumerated().map { index, color in
      Gradient.Stop(color: color, location: locations[safe: index] ?? fallbackLocation(index))
    }
  }

  private func fallbackLocation(_ index: Int) -> Double {
    colors.count <= 1 ? 0 : Double(index) / Double(colors.count - 1)
  }
}

struct MossyUnevenCornerRadiusModifier: ViewModifier, Record {
  @Field var topLeft: CGFloat?
  @Field var topRight: CGFloat?
  @Field var bottomRight: CGFloat?
  @Field var bottomLeft: CGFloat?

  func body(content: Content) -> some View {
    content.clipShape(
      MossyUnevenRoundedRectangle(
        topLeft: topLeft ?? 0,
        topRight: topRight ?? 0,
        bottomRight: bottomRight ?? 0,
        bottomLeft: bottomLeft ?? 0
      )
    )
  }
}

struct MossyDirectionalBorderModifier: ViewModifier, Record {
  @Field var color: Color = .clear
  @Field var top: CGFloat?
  @Field var right: CGFloat?
  @Field var bottom: CGFloat?
  @Field var left: CGFloat?

  func body(content: Content) -> some View {
    content.overlay {
      GeometryReader { geometry in
        let size = geometry.size
        ZStack {
          if let top, top > 0 {
            Rectangle()
              .fill(color)
              .frame(width: size.width, height: top)
              .position(x: size.width / 2, y: top / 2)
          }
          if let right, right > 0 {
            Rectangle()
              .fill(color)
              .frame(width: right, height: size.height)
              .position(x: size.width - right / 2, y: size.height / 2)
          }
          if let bottom, bottom > 0 {
            Rectangle()
              .fill(color)
              .frame(width: size.width, height: bottom)
              .position(x: size.width / 2, y: size.height - bottom / 2)
          }
          if let left, left > 0 {
            Rectangle()
              .fill(color)
              .frame(width: left, height: size.height)
              .position(x: left / 2, y: size.height / 2)
          }
        }
      }
      .allowsHitTesting(false)
    }
  }
}

struct MossyUnevenRoundedRectangle: Shape {
  var topLeft: CGFloat
  var topRight: CGFloat
  var bottomRight: CGFloat
  var bottomLeft: CGFloat

  func path(in rect: CGRect) -> Path {
    let topLeft = min(topLeft, min(rect.width, rect.height) / 2)
    let topRight = min(topRight, min(rect.width, rect.height) / 2)
    let bottomRight = min(bottomRight, min(rect.width, rect.height) / 2)
    let bottomLeft = min(bottomLeft, min(rect.width, rect.height) / 2)

    var path = Path()
    path.move(to: CGPoint(x: rect.minX + topLeft, y: rect.minY))
    path.addLine(to: CGPoint(x: rect.maxX - topRight, y: rect.minY))
    path.addArc(
      center: CGPoint(x: rect.maxX - topRight, y: rect.minY + topRight),
      radius: topRight,
      startAngle: .degrees(-90),
      endAngle: .degrees(0),
      clockwise: false
    )
    path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY - bottomRight))
    path.addArc(
      center: CGPoint(x: rect.maxX - bottomRight, y: rect.maxY - bottomRight),
      radius: bottomRight,
      startAngle: .degrees(0),
      endAngle: .degrees(90),
      clockwise: false
    )
    path.addLine(to: CGPoint(x: rect.minX + bottomLeft, y: rect.maxY))
    path.addArc(
      center: CGPoint(x: rect.minX + bottomLeft, y: rect.maxY - bottomLeft),
      radius: bottomLeft,
      startAngle: .degrees(90),
      endAngle: .degrees(180),
      clockwise: false
    )
    path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + topLeft))
    path.addArc(
      center: CGPoint(x: rect.minX + topLeft, y: rect.minY + topLeft),
      radius: topLeft,
      startAngle: .degrees(180),
      endAngle: .degrees(270),
      clockwise: false
    )
    path.closeSubpath()
    return path
  }
}

extension Array {
  subscript(safe index: Int) -> Element? {
    indices.contains(index) ? self[index] : nil
  }
}

File: android/src/main/java/com/mossyui/MossyUiModule.kt
package com.mossyui

import android.graphics.Color as AndroidColor
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.ui.ModifierRegistry

class MossyUiModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MossyUi")

    OnCreate {
      ModifierRegistry.register("mossySizeConstraints") { map, _, _, _ ->
        Modifier.sizeIn(
          minWidth = map.dp("minWidth") ?: 0.dp,
          minHeight = map.dp("minHeight") ?: 0.dp,
          maxWidth = map.dp("maxWidth") ?: androidx.compose.ui.unit.Dp.Infinity,
          maxHeight = map.dp("maxHeight") ?: androidx.compose.ui.unit.Dp.Infinity
        )
      }

      ModifierRegistry.register("mossyLinearGradientBackground") { map, _, _, _ ->
        val colors = map.colors("colors")
        if (colors.isEmpty()) {
          Modifier
        } else {
          val locations = map.floatList("locations")
          val colorStops = colors.mapIndexed { index, color ->
            (locations.getOrNull(index) ?: fallbackLocation(index, colors.size)) to color
          }.toTypedArray()
          val startPoint = map.point("startPoint", Offset.Zero)
          val endPoint = map.point("endPoint", Offset(1f, 1f))

          Modifier.drawBehind {
            drawRect(
              brush = Brush.linearGradient(
                colorStops = colorStops,
                start = Offset(size.width * startPoint.x, size.height * startPoint.y),
                end = Offset(size.width * endPoint.x, size.height * endPoint.y)
              )
            )
          }
        }
      }

      ModifierRegistry.register("mossyUnevenCornerRadius") { map, _, _, _ ->
        Modifier.clip(
          RoundedCornerShape(
            topStart = (map.dp("topLeft") ?: 0.dp),
            topEnd = (map.dp("topRight") ?: 0.dp),
            bottomEnd = (map.dp("bottomRight") ?: 0.dp),
            bottomStart = (map.dp("bottomLeft") ?: 0.dp)
          )
        )
      }

      ModifierRegistry.register("mossyDirectionalBorder") { map, _, _, _ ->
        val color = map.color("color") ?: return@register Modifier
        Modifier.drawBehind {
          val top = map.dp("top")?.toPx() ?: 0f
          val right = map.dp("right")?.toPx() ?: 0f
          val bottom = map.dp("bottom")?.toPx() ?: 0f
          val left = map.dp("left")?.toPx() ?: 0f

          if (top > 0f) {
            drawLine(color, Offset(0f, top / 2), Offset(size.width, top / 2), strokeWidth = top)
          }
          if (right > 0f) {
            drawLine(
              color,
              Offset(size.width - right / 2, 0f),
              Offset(size.width - right / 2, size.height),
              strokeWidth = right
            )
          }
          if (bottom > 0f) {
            drawLine(
              color,
              Offset(0f, size.height - bottom / 2),
              Offset(size.width, size.height - bottom / 2),
              strokeWidth = bottom
            )
          }
          if (left > 0f) {
            drawLine(color, Offset(left / 2, 0f), Offset(left / 2, size.height), strokeWidth = left)
          }
        }
      }
    }

    OnDestroy {
      ModifierRegistry.unregister("mossySizeConstraints")
      ModifierRegistry.unregister("mossyLinearGradientBackground")
      ModifierRegistry.unregister("mossyUnevenCornerRadius")
      ModifierRegistry.unregister("mossyDirectionalBorder")
    }
  }
}

private fun Map<String, Any?>.number(key: String): Float? = (this[key] as? Number)?.toFloat()

private fun Map<String, Any?>.dp(key: String) = number(key)?.dp

private fun Map<String, Any?>.floatList(key: String): List<Float> =
  (this[key] as? List<*>)?.mapNotNull { (it as? Number)?.toFloat() }.orEmpty()

private fun Map<String, Any?>.colors(key: String): List<Color> =
  (this[key] as? List<*>)?.mapNotNull { (it as? String)?.toComposeColor() }.orEmpty()

private fun Map<String, Any?>.color(key: String): Color? = (this[key] as? String)?.toComposeColor()

private fun Map<String, Any?>.point(key: String, fallback: Offset): Offset {
  val value = this[key] as? Map<*, *> ?: return fallback
  val x = (value["x"] as? Number)?.toFloat() ?: fallback.x
  val y = (value["y"] as? Number)?.toFloat() ?: fallback.y
  return Offset(x, y)
}

private fun String.toComposeColor(): Color? = runCatching {
  Color(AndroidColor.parseColor(expandShortHexColor()))
}.getOrNull()

private fun String.expandShortHexColor(): String {
  if (!startsWith("#")) return this

  return when (length) {
    4 -> "#${this[1]}${this[1]}${this[2]}${this[2]}${this[3]}${this[3]}"
    5 -> "#${this[4]}${this[4]}${this[1]}${this[1]}${this[2]}${this[2]}${this[3]}${this[3]}"
    else -> this
  }
}

private fun fallbackLocation(index: Int, count: Int): Float {
  return if (count <= 1) 0f else index.toFloat() / (count - 1).toFloat()
}
